elasticdl/python/ps/servicer.py

```diff
@@ -25,8 +26,31 @@ def __init__(
         self._lock = threading.Lock()
 
     def pull_variable(self, request, _):
-        # TODO: implement this RPC service
-        return elasticdl_pb2.PullVariableResponse()
+        """
+        Response with all non-embedding parameters if initialized.
+        """
+        res = elasticdl_pb2.PullVariableResponse()
+        if not self._parameters.init_status:
+            res.model_init_status = False
+            return res
+
+        # Only sync-SGD needs lock
+        # TODO: use a read-write lock to support multiple concurrent reads
```

Comment:
What does this TODO means? There is already a lock when using sync-SGD.

Reply:
In sync-SGD, if PS is updating parameters with gradients, it should hold on the pull_variable from workers until the updates are done. Otherwise, workers may get inconsistent model variables.

Reply:
Got it. How about modify the TODO to "If no updates happening, do not lock and support concurrent reads"?

Reply:
If no update happening, the lock is still needed to prevent updates in the middle of this read service.
A read-write lock is needed. If no write, multiple reads. If write needed, all new reads are blocked,  wait for all existing reads done, then write.